credentials/credentials_test.go

```diff
@@ -58,4 +62,173 @@ func TestTLSClone(t *testing.T) {
 	if c.Info().ServerName != expectedServerName {
 		t.Fatalf("Change in clone should not affect the original, c.Info().ServerName = %v, want %v", c.Info().ServerName, expectedServerName)
 	}
+
+}
+
+const tlsDir = "../test/testdata/"
+
+type serverHandshake func(net.Conn, *tls.ConnectionState) error
+
+func TestClientHandshakeReturnsAuthInfo(t *testing.T) {
+	var serverConnState tls.ConnectionState
+	errChan := make(chan error, 1)
+	lisAddr, err := launchServer(t, &serverConnState, tlsServerHandshake, errChan)
+	if err != nil {
+		return
+	}
+	clientConnState, err := clientHandle(t, gRPCClientHandshake, lisAddr)
+	if err != nil {
+		return
+	}
+	// wait until server has populated the serverAuthInfo struct or failed.
+	if err = <-errChan; err != nil {
+		return
+	}
+	if !isEqualState(clientConnState, serverConnState) {
+		t.Fatalf("c.ClientHandshake(_, %v, _) = %v, want %v.", lisAddr, clientConnState, serverConnState)
+	}
+}
+
+func TestServerHandshakeReturnsAuthInfo(t *testing.T) {
+	var serverConnState tls.ConnectionState
+	errChan := make(chan error, 1)
+	lisAddr, err := launchServer(t, &serverConnState, gRPCServerHandshake, errChan)
+	if err != nil {
+		return
+	}
+	clientConnState, err := clientHandle(t, tlsClientHandshake, lisAddr)
+	if err != nil {
+		return
+	}
+	// wait until server has populated the serverAuthInfo struct or failed.
+	if err = <-errChan; err != nil {
+		return
+	}
+	if !isEqualState(clientConnState, serverConnState) {
+		t.Fatalf("ServerHandshake(_) = %v, want %v.", serverConnState, clientConnState)
+	}
+}
+
+func TestServerAndClientHandshake(t *testing.T) {
+	var serverConnState tls.ConnectionState
+	errChan := make(chan error, 1)
+	lisAddr, err := launchServer(t, &serverConnState, gRPCServerHandshake, errChan)
+	if err != nil {
+		return
+	}
+	clientConnState, err := clientHandle(t, gRPCClientHandshake, lisAddr)
+	if err != nil {
+		return
+	}
+	// wait until server has populated the serverAuthInfo struct or failed.
+	if err = <-errChan; err != nil {
+		return
+	}
+	if !isEqualState(clientConnState, serverConnState) {
+		t.Fatalf("Connection states returened by server: %v and client: %v aren't same", serverConnState, clientConnState)
+	}
+}
+
+func isEqualState(state1, state2 tls.ConnectionState) bool {
+	if state1.Version == state2.Version &&
+		state1.HandshakeComplete == state2.HandshakeComplete &&
+		state1.CipherSuite == state2.CipherSuite &&
+		state1.NegotiatedProtocol == state2.NegotiatedProtocol {
+		return true
+	}
+	return false
+}
+
+func launchServer(t *testing.T, serverConnState *tls.ConnectionState, hs serverHandshake, errChan chan error) (string, error) {
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Errorf("Failed to listen: %v", err)
```

Comment:
you can t.Fatalf directly so that this function only needs to return lis.Addr().String().